x/upgrade/types/handler.go

```diff
@@ -2,7 +2,8 @@ package types
 
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
 )
 
 // UpgradeHandler specifies the type of function that is called when an upgrade is applied
-type UpgradeHandler func(ctx sdk.Context, plan Plan)
+type UpgradeHandler func(ctx sdk.Context, plan Plan, versionMap module.VersionMap) error
```

Comment:
I guess my original idea was that we don't pass a reference to `ModuleManager` into `x/upgrade` but instead communicate the new versions with the return type... However, to avoid bike-shedding let's just go with this approach for now as I think they're basically equivalent.

Reply:
Ah... actually I think this is an issue. So the code in this PR fails if the module manager is not passed in. We should either:
* fail gracefully if the module manager is not set, or
* take the alternate approach I'm suggesting

Reply:
I like the Aarons approach. Limiting dependencies (especially for something as big as `ModuleManager`) sounds as a good approach.

Reply:
Re:
```
err := handler(ctx, plan, versionMapFromState) // Run the handler with module versions from state k.SetCurrentConsensusVersion(ctx); // Once the handler is run, update state with new module versions
```
Communicating through context smells bad - we obfuscate the information.

Reply:
@technicallyty @amaurym what if went with my proposed method signature above and instead of `SetVersionManager` we have a method `SetInitialVersions` so that we're not passing a reference to the module manager directly into the upgrade keeper? (Sorry for the back and forth on all this @technicallyty 😉 )

Reply:
OK sgtm, we can do like this 👍

Reply:
What if we just avoid the extra field altogether, export the `setConsensusVersions` function and update it to take a `VersionMap`? The `versionMap` field on the keeper only ever gets used once anyways, which is from the `InitGenesis` call where we set the initial consensus versions to state. If we go with my proposed change, we don't need the extra field on upgrade's keeper, nor the InitGenesis code. Are there any advantages to committing the versions to state through InitGenesis rather than setting it outright at the end of simapp?

We add this line to the bottom of simapp
```diff
ctx := app.BaseApp.NewUncachedContext(true, tmproto.Header{})
app.CapabilityKeeper.InitializeAndSeal(ctx)
+ app.UpgradeKeeper.SetConsensusVersions(ctx, app.mm.GetVersionMap())
```

thoughts @amaurym, @robert-zaremba, @aaronc ?

Reply:
> Are there any advantages to committing the versions to state through InitGenesis rather than outright at the end of simapp?

End of simapp would not work in the following scenario:
- old binary runs, saves old VersionMap in x/upgrade store
- swap binary
- new binary runs. It doesn't run InitGenesis (it's not a new chain), but it runs "end of simapp", so saves  app.mm.VersionMap to x/upgrade store
- after that, it runs the UpgradeHandler. But it detects that the Versionmap inside x/upgrade store is exactly the same as the app.mm.VersionMap one, so it doesn't run any migration.

Does that make sense?

We still need to call `UpgradeKeeper.SetConsensusVersions` twice:
- once in InitGenesis
- a second time inside the UpgradeHandler

Reply:
Definitely true what you're saying about simapp @amaurym.

Here's what I suggest:
* have a `UpgradeKeeper.SetConsensusVersions(VersionMap)` method - this sets current versions which get written during `InitChain`. So in `app.go`, users will write:
```go
upgradeKeeper.SetConsensusVersions(mm.GetConsensusVersions())
```
* add a `VersionMap` return parameter to `RunMigrations` and a ` VersionMap` return parameter to `Upgrade Handler`, so that most upgrade handlers will simply be implemented as:
```go
func(ctx sdk.Context, plan Plan, versionMap module.VersionMap) (VersionMap, error) {
  return mm.RunMigrations(ctx, versionMap)
}
```

Unless there is a strong objection, I suggest we go with this. And if we need to go into any more details, I suggest we get on a short call to reach consensus.

Reply:
Makes sense, [changes implemented](https://github.com/cosmos/cosmos-sdk/pull/8743/commits/799b646a97f46c6e2d1c8fbf727806b11c91d866)